File: static/metaDefinition/pciProperties.js
/* 
 * Definicion de las propiedades, 
 * 
 * las propiedades tienen 
 * 
 *      help         : description 
 *      type    
 *      choices      : lista de valores en caso de ser string 
 *      required     : 
 * 
 *  EL nombre de la propiedad contiene el defaultValue q solo se aplcia 
 *  si no existe y es requerida  
 * 
 */

// La defincion de tipos es   xxx.type = [ 'boolean' | 'date' | 'string' | 'number' ] 


function verifyPrpType(  lKey,  cValue ) {
    /* Verifica los tipos de las  propiedades
     * recibe el valor y el tipo y verifica si 
     * corresponden entre si
     * 
     * Intenta la conversion, sin no regresa nulo 
     */ 
    
    var pType = _MetaProperties[ lKey + '.type' ] 
    if ( ! pType )  { 
        if ( typeof ( cValue ) == 'string') { return cValue.trimRight()  }
        else { return cValue }      
    }
    
    if ( pType == typeof( cValue ) ) 
    { return cValue  }    
    
    switch ( pType ) {
    case "boolean":
        if  ( (typeof( cValue ) == 'number') ){ cValue = cValue.toString() }
        if  ( (typeof( cValue ) == 'string') ){
            if ( cValue.substr(1,1).toLowerCase()  in oc([ 'y', 's', '1','o' ]) )
            { return true } else { return false } 
        } else { return false }  
    case "number":
        return parseFloat( cValue );
    case "null":
        return null 
    default:
        return cValue
    }
}



_MetaProperties =  {

    "metaVersion.help" : "Version interna de la meta", 
    "userVersion.help" : "Version de usuario ( editable )",
     
    "exportCsv.help"   : "Permite exportar la definicion de fields en formato csv",   
    "exportCsv.type" : "boolean", 

    
//  Types
    "required.type" : "boolean", 
    "fromModel.type"  : "boolean",
    "hideRowNumbers.type"  : "boolean",
    "multiSelect.type"  : "boolean",
    "readOnly.type"  : "boolean",
    
// PCI
    "protoConcept.help" : "Opcion de base de la pci (apunta directamente al modelo Django  app.modelo)",
    "protoOption.help"  : "Definicion de view interface - puede ser = al protoConcpeto o definir un nivel adicional app.modelo.vista", 
    "description.help"  : "Text descriptivo", 
    "protoIcon.help"  : "iconName tal como se define en el css",
    "shortTitle.help" : "Titulo del menu y de la forma" ,
    "idProperty.help" : "Campo q sirve como Id en el modelo, Django definie por defecto un Id, no es necesario definirlo explitamente",
     
    "pciStyle" : "grid", 
    "pciStyle.help" : "Presentacion de la pci [ form,  grid, tree]", 
    "pciStyle.choices": ["grid", "form", "tree"],

    "denyAutoPrint.type" : "boolean", 
    "denyAutoPrint.help" : "Impide la impresion automatica de la grilla", 

// Detalles 
    "masterField" : "pk",

    "menuText.help"         : "titulo en el menu ( toolbar )", 
    "detailName.help"       : "key del detalle usada para encadenar los reportes ", 
    
    "conceptDetail.help"    : "Entidad detalle de la relacion ( [App.]Model o [App.]Model.View )",
    "masterField.help"      : "Campo en el maestro para q contiene el criterio para el filtro del detalle, (normalmente Pk si es un detalle directo)",
    "detailField.help"      : "Campo del detalle para hacer el filtro, normalmente el nombre de la tabla padre en minuscula", 
    "detailTitleLbl.help"   : "Titulo en el detalle para indicar el filtro actual", 

    "masterTitleField.help" : "Campo en el maestro que se utiliza para construir el titulo de la vista filtrada en el detalle", 
    "detailTitleField.help" : "Campo en el detalle donde se copia el valor del titulo del maestro ( default de la llave padre en la edicion )", 

//  Udps 
    "propertyPrefix" : "udp", 
    "propertyPrefix.help" : "prefijo de las propiedades definidas por el usuario ( en el nombre del campo aparencen upd__xxxxx)", 
    "propertyName.help" : "campo que contiene la llave de la propiedad",
    "propertyValue.help" : "campo que contiene el valor",

    "propertyRef.help" : "REQUERIDO: Campo en la udp q apunta a la tabla base",    
    "keyField.help" : "Campo leido del registro de base<br>** Solo se debe setear cuando la udp no es un MD",
    "udpTable.help" : "Tabla que contiene las upds, <b>** Si es un vinculo directo corresponde al related_name q es el set de detalles, normalmente la tablaUdp comenzando por minuscula", 

// sheets
    "protoSheetSelector.help": "Campo de criterio para seleccionar el template, vacio para una unica plantilla por defecto DEFAULT",
    "template.help": "Definicion de la plantilla", 

    "templateFp.help": "Templante FirstPage", 
    "templateLp.help": "Templante LastPage",
     
    "templateBb.help": "Templante BeforeBlock", 
    "templateAb.help": "Templante AfterBlock", 

    "templateBb": "<spam>------------------------------------</spam><br>", 
    "templateAb": "<spam>====================================</spam><br>", 

    "templateEr.help": "EveryRow", 

    "templateFp": '<!DOCTYPE html>' + 
        '<html>' +
          '<head>' +
           '<meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />' +
           '<link href="/static/css/print.css" rel="stylesheet" type="text/css" media="screen,print" />' +
           '<title>PtReport : {{reportTitle}}</title>' +
          '</head>' +
        '<body>',  

    "templateLp": '</body>' +
                  '</html>',  


// 
    "required": false,
    "required.help": "Requiere valores la forma",
    "allowDecimals.help": "NO USAR : si permite o no decimales, si decimalPReciosion = 0 implica falso",
    "autoscroll": true,
    "autoscroll.help": "t/f",

    "choices.help": "lista de valores separados por coma para el combobox",
    
    "collapsed": false,
    "collapsed.help": "t/f aparece contraido",
    "collapsed.type": "boolean",

    "collapsible": false,
    "collapsible.help": "t/f",
    "collapsible.type": "boolean",

    "cellToolTip.help": "Presenta contenido del campo como micro ayuda",
    "cellToolTip.type": "boolean",

    "cellLink.help": "Presenta contenido del campo como micro ayuda",
    "cellLink.type": "boolean",

    "fromModel.help" : "Pertenece al modelo de base",  
    "xtype.help" : "Tipo de objeto en el frontEnd ( puede ser manipulado con el vType)",
    "xtype.choices": ["", "textfield", "combobox", "checkbox", "numberfield", "textarea", "datefield" ],
    
    "decimalPrecision": 0,
    "decimalPrecision.help": "Cantidad de decimales permitidos ( 0 para enteros 2 para decimales )",
    "defaultValue.help": "valor por defecto",
    
    "fieldLabel.help": "label en la forma",
    "format.help": "mascara del campo ( automatico para las fechas, horas y numeros )",
    "fsLayout": "fluid",
    "fsLayout.choices": ["fluid", "1col", "2col", "3col" ],
    "fsLayout.help": "distribucion de los campos en el contenedor",

    "header.help" : "Encabezado de la collumna en modo grilla ( default para fieldLabel ) " ,

    "height.type": "number",
    "height.help": "The height value in pixels",
    "helpPath.help": "se define como : /help/xxxx.html  esta en: /static/help",
    
    "hidden.help": "componente oculto",
    "hidden.type": "boolean",

    "hideLabel.help": "presenta los campos sin label, util cuando se configura grupos como firstName, lastName",
    "hideSheet.help" : "Oculta la ficha descriptiva", 
    "hideRowNumbers.help" : "Oculta la columna que numera los campos", 
    "multiSelect.help" : "Permite seleccionar multiples registros", 

    "filterSetABC.help" : "Genera automaticamente filtro alfabetico sobre la columna seleccionada", 

    "labelAlign": "left",
    "labelAlign.choices": ["left", "top"],
    "labelAlign.help": "opciones left top",

    "labelWidth.help": "ancho del label",

    "maxHeight.help": "The max value in pixels",
    "maxWidth.help": "The max value in pixels",

    "maxHeight.type": "number",
    "maxWidth.type": "number",

    "minHeight.help": "The minimum value in pixels",
    "minWidth.help": "The minimum value in pixels", 

    "minHeight.type": "number",
    "minWidth.type": "number", 

    "height.type": "number",
    "width.type": "number", 

    "flex.type": "number", 
    "flex.help": "Recalcula el ancho en funcion de la forma y el peso ( flex) asignado", 

    "readOnly": false,
    "readOnly.help": "campo de solo lectura",
    "title.help": "Titulo del componente",
    "tooltip.help": "microayuda del componente ",
 
    "sortable.help": "t/f",
    "sortable.type": "boolean",

    "type.help" : "Tipo de dato de la Db ", 
    "type.choices" : [ "", "string", "text", "bool", "int", "decimal", "combo", "date",  "datetime", "time", "autofield", "html", "foreignid",  "foreigntext"  ],             

    "vType.help" : "TODO: Tipo de dato de validacion  ", 
    "vType.choices" : [ "", "email", "ip4", "ip6", "tel", "postalCodeCA", "postalCodeUSA"  ],             
    
    "width.help": "The width value in pixels",
    
    "wordWrap.type": "boolean",
    "wordWrap.help": "ver el contenido en mas de una linea", 

    "sheetType.help" : "Tipo de hoja ( usada para reportes ) ", 
    "sheetType.choices" : [ "", "printerOnly", "gridOnly"], 
    
    "detailName.help" : "Nombre correspondiente a un detalle declarado en la opcionBase correspondiente ( padre )",              

    "actionType.help" : "Tipo de accion ( metodos backend ) ", 
    "actionType.choices" : [ "", "report", "wflow", "edicion"], 

    "refreshOnComplete.type": "boolean",
    
    "actionParamType.help" : "Tipo de dato del parametro", 
    "actionParamType.choices" : [ "", "string", "bool", "number"], 
    
    "fromField.help" : "Copia el contenido de un campo en otro, en el caso de zooms o valores por defecto" 
    

};

File: mariane/protoPci/pciProperties.py
import json
# Definicion de las propiedades, 
# las propiedades tienen 
#      help         : description 
#      type    
#      choices      : lista de valores en caso de ser string 
#      required     : 
# 
#  EL nombre de la propiedad contiene el prpDefault q solo se aplcia 
#  si no existe y es requerida  


# La defincion de tipos es   xxx.type = [ 'boolean' | 'date' | 'string' | 'number' ] 



_MetaProperties =  {

    "metaVersion.help" : "Internal meta version", 
    "userVersion.help" : "Application version",
     
    "exportCsv.help"   : "Csv export enabled?",   
    "exportCsv.type" : bool, 

    "localSort.help" : "local sort?, (n/a for prototypes)", 
    "localSorttype" : bool,
     
    "pageSize,help" : "Page size", 
    "pageSize.type": int, 

#QBE    
    "qbeHelp.type": bool,
    "qbeHelp.help": "visible trigger for select distinct (interne)",

#  Types
    "required.type" : bool, 
    "readOnly.type"  : bool,
    "primary.type" : bool, 
    
# PCI
    "viewEntity.help" : "Backend model (Django)",
    "viewCode.help"  : "Definicion de view code -  app.model.view", 
    "description.help"  : "Description", 
    "viewIcon.help"  : "iconName  (css name)",
    "shortTitle.help" : "Menu and form title" ,
    "idProperty.help" : "Id property (future use)",

    "protoEntity.help" : "Default prototype entity ( prototype.protoTable.xxx )  (Internal use with protoEntityId)",
     
    "pciStyle" : "grid", 
    "pciStyle.help" : "Presentation mode [ form,  grid, tree]", 
    "pciStyle.choices": ["grid", "form", "tree"],

    "gridSelectionMode.choices": ["multi", "simple", "single" ], 
    "gridSelectionMode.help":  "multi*: multiple selection with check; simple: selection on/off ; single: Last selected", 

    "denyAutoprint.type" : bool, 
    "denyAutoprint.help" : "deny auto print (future use)", 

# Detalles 
    "masterField" : "pk",

    "menuText.help"         : "Menu title ( toolbar )", 
    "detailName.help"       : "Detail key (for report detail)", 
    
    "conceptDetail.help"    : "Detail concept ( [App.]Model o [App.]Model.View )",
    "masterField.help"      : "Master field for MD navigation (normaly Pk)",
    "detailField.help"      : "Detail field for filter (normaly conceptName)", 
    "detailTitleLbl.help"   : "Detail title for filter", 

    "masterTitleField.help" : "Master field title (filter name)", 
    "detailTitleField.help" : "Master field title (copy from master title in detail edition)", 

#  Udps 
    "propertyPrefix" : "udp", 
    "propertyPrefix.help" : "udp prefix (upd__xxxxx)", 
    "propertyName.help" : "udp property name",
    "propertyValue.help" : "udp property value",

    "propertyRef.help" : "udp concept ref",    
    "keyField.help" : "udp source<br>** Only for non MD udp (internal use)",
    "udpTable.help" : "udp concept name , <strong>** if direct link then related_name, normaly udpTable</strong>", 

# sheets
    "sheetSelector.help": "field sheet selecto, null for DEFAULT sheet",
    "template.help": "template definition", 

    "templateFp.help": "Templante FirstPage", 
    "templateLp.help": "Templante LastPage",
     
    "templateBb.help": "Templante BeforeBlock", 
    "templateAb.help": "Templante AfterBlock", 

    "templateBb": "<spam>------------------------------------</spam><br>", 
    "templateAb": "<spam>====================================</spam><br>", 

    "templateEr.help": "EveryRow", 

    "templateFp":'<!DOCTYPE html><html><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type" /><link href="/static/css/print.css" rel="stylesheet" type="text/css" media="screen,print" /><title>PtReport : {{reportTitle}}</title></head><body>',
    "templateLp":'</body></html>',

    "direction": "sort order",
    "direction.choices": ["ASC", "DESC" ],

#  Fields
    "physicalName.help" : "phisical name or function  @str( f1,f2 )", 
    "required": False,
    "required.help": "Required field",
    "allowDecimals.help": "Dont use!!! : allow decimal (internal use)",
    "autoscroll": True,
    "autoscroll.help": "t/f",

    "choices.help": "comma separed values for combo selection",
    
    "collapsed": False,
    "collapsed.help": "t/f collapsed",
    "collapsed.type": bool,

    "zoomMultiple": False, 
    "zoomMultiple.type": bool, 
    "zoomMultiple.help": "Multiple selection on add",

    "collapsible": False,
    "collapsible.help": "t/f",
    "collapsible.type": bool,

    "cellToolTip.help": "Use this field form line tooltip",
    "cellToolTip.type": bool,

    "cellLink.help": "Is a cellLink?",
    "cellLink.type": bool,

    "xtype.help" : "frontend widget (use vType also)",
    "xtype.choices": ["", "textfield", "combobox", "checkbox", "numberfield", "textarea", "datefield" ],
    
    "prpScale": 0,
    "prpScale.help": "Decimal scale ( 0 int, 2 dec )",
    "prpDefault.help": "Default value",
    
    "fieldLabel.help": "Field label (in form)",
    "format.help": "input mask (automatic for date, time and numbers) (future use)",
    "fsLayout": "fluid",
    "fsLayout.choices": ["fluid", "1col", "2col", "3col" ],
    "fsLayout.help": "Automatic layout distribution",

    "header.help" : "Column header (grid)  default for fieldLabel" ,

    "height.type": int,
    "height.help": "The height value in pixels",
    "helpPath.help": "/help/xxxx.html  real: /static/help",
    
    "hidden.help": "Hiden field (future use)",
    "hidden.type": bool,

    "hideLabel.help": "Hide label? (ex: firstName, lastName)",
    "hideSheet.help" : "Hide sheet?", 
    
    "hideRowNumbers.help" : "Hide row numbers?", 
    "hideRowNumbers.type"  : bool,
    
    "hideCheckSelect.type"  : bool,
    "hideCheckSelect.help" : "Hide check select?", 

    "filterSetABC.help" : "Auto alphabetic filter", 

    "labelAlign": "left",
    "labelAlign.choices": ["left", "top"],
    "labelAlign.help": "Label align (left, top)",

    "labelWidth.help": "Label width",

    "maxHeight.help": "The max value in pixels",
    "maxWidth.help": "The max value in pixels",

    "maxHeight.type": int,
    "maxWidth.type": int,

    "minHeight.help": "The minimum value in pixels",
    "minWidth.help": "The minimum value in pixels", 

    "minHeight.type": int,
    "minWidth.type": int, 

    "height.type": int,
    "width.type": int, 

    "flex.type": int, 
    "flex.help": "Flex width eqivalence", 

    "readOnly": False,
    "readOnly.help": "ReadOnly field?",
    "title.help": "Title",
    "tooltip.help": "Microhelp",
 
    "sortable.help": "Sortable?",
    "sortable.type": bool,

    "searchable.help": "Searchable?",
    "searchable.type": "bool",

    "type.help" : "Field type", 
    "type.choices" : [ "", "string", "text", "bool", "int", "decimal", "combo", "date",  "datetime", "time", "autofield", "html", "foreignid",  "foreigntext"  ],             

    "vType.help" : "Validation type", 
    "vType.choices" : [ "", "email", "ip4", "ip6", "tel", "postalCodeCA", "postalCodeUSA"  ],             
    
    "width.help": "The width value in pixels",
    
    "wordWrap.type": bool,
    "wordWrap.help": "Auto wordWrap (more than one line)", 

    "sheetType.help" : "Sheet type (Grid, Report)", 
    "sheetType.choices" : [ "", "printerOnly", "gridOnly"], 
    
    "detailName.help" : "Detail name (used in MD definition)",              

    "actionType.help" : "Action type (backend function)", 
    # Solo las acciones de tipo "user" son presentadas  en el menu
    # los triggers pueden ser escritos directamente en el modelo y no pasar por protoExt 
    "actionType.choices" : [ "user", "insTrigger", "updTrigger", "delTrigger", "wflow"], 
    "refreshOnComplete.type": bool,
    
    "paramType.help" : "Parameter type", 
    "paramType.choices" : [ "", "string", "bool", "number"], 

    "cpFromField.help" : "Derived by copy",  
    "cpFromZoom.help" : "Derived form zoom (fk property)? ", 
    
    "crudType.help" : "Crud type", 
    # editable      : es un campo estandar de la Db ( default )  
    # screenOnly    : ninguna iteraccion con la db, funciones calculadas en el frontEnd, o campos de procesamiento intermedio para generar otros campos     
    # storeOnly     : leido de la Db,  no se despliega en el frontEnd, se usa como resultado de campos calculados, usado para manejar subSets ( implica definir baseFilter, vrDefault  )
    # insertOnly    : campos invariables ( ej: nro documento, )      
    # updateOnly    : nulo al inicio, requerido en modificacion       
        
    # linked        : no es editable, no se guarda en la Db, requiere cpFromField,  cpFromZoom* ( *para prototipos, o zooms no relacionales )  
    # copied        : toma el vr por defecto de cpFromField o cpFromZoom ( similar a linked + editable )      
    "crudType.choices" : [ "", "editable", "screenOnly", "storeOnly", "insertOnly", "updateOnly", "linked", "copied" ], 
    
    "selectionMode.help" : "Grid selection mode for actions",
    # none : Envia la accion sin QSet 
    # single : Exige un unico reg 
    # multiple : Exige al menos un reg       
    "selectionMode.choices" : [ "none", "single", "multiple" ] 

};

def verifyPrpType(lKey, cValue) :
    # Verifica los tipos de las  propiedades
    # recibe el valor y el tipo y verifica si 
    # corresponden entre si
    # Intenta la conversion, sin no regresa nulo  

    pType = _MetaProperties.get(lKey + '.type')    
    if(not pType):
        if (type(cValue) == str):
            return cValue.replace('~+$', '')
        
        else :
            return cValue 

    
    if (pType == type(cValue)) :
        return cValue  
    
    elif(pType == bool):
        if((type( cValue ) == int) or (type( cValue ) == float)) :
            cValue = str(cValue)
             
        if((type( cValue ) == str) ) :
            if (cValue[0:2].lower() in [ 'y', 's', '1','o', 't' ]):
                return True
            else :
                return False
                 
        else :
            return False
        
    elif(pType == int ):
        return int(cValue)
    
    elif(pType == float):
        return float(cValue)
    
    elif(pType == None):
        return None
    
    else :
        return cValue;



def getSimpleProperties(oData, ptType) :
    # Retorna los valores simples, verificando los tipos de cada propiedad

    # Solo deben llegar objetos, si llega un array no hay props q mostrar
    if(type(oData) == list) :
        return [];

    # Inicializa con el type
    cData = {}
    if(ptType) :
        cData['__ptType'] = ptType;

    for lKey in oData :
        cValue = oData[lKey];
 
        # Los objetos o arrays son la imagen del arbol y no deben ser tenidos en cuenta, generarian recursividad infinita
        if (type(cValue) in [dict, list]) :
            continue

        # Si son valores codificados, los decodifica y los agrega
        if (lKey in ['__ptValue', '__ptList']) :
            try :

                cData = json.loads(cValue)
                        
            except :

                #print( "Error de encodage", cValue )
                pass
                    
        else :
            cValue = verifyPrpType(lKey, cValue);
            if (cValue) :
                if (type(cData)== dict):
                    cData[lKey] = cValue
                else :
                    cData += {lKey : cValue}
                        
    return cData 
